src/main/java/bio/terra/service/dataset/DatasetDao.java

```diff
@@ -114,6 +115,73 @@ public boolean unlock(UUID datasetId, String flightId) {
         return (numRowsUpdated == 1);
     }
 
+    /**
+     * Take a shared lock on the dataset object before doing something with it (e.g. file ingest, file delete).
+     * This method returns successfully when this flight has a shared lock on the dataset object, and
+     * throws an exception in all other cases. So, multiple locks can succeed with no errors. Logic flow of the method:
+     *     1. Update the dataset record to give this flight a shared lock.
+     *     2. Throw an exception if no records were updated.
+     * @param datasetId id of the dataset to lock, this is a unique column
+     * @param flightId flight id that wants to lock the dataset
+     * @throws DatasetLockException if another flight has an exclusive lock on the dataset
+     * @throws DatasetNotFoundException if the dataset does not exist
+     */
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void lockShared(UUID datasetId, String flightId) {
+        if (flightId == null) {
+            throw new DatasetLockException("Locking flight id cannot be null");
+        }
+
+        // update the dataset entry and lock it by adding the flight id to the shared lock array column
+        String sql = "UPDATE dataset SET sharedlock = (CASE " +
+            "WHEN sharedlock IS NULL " +
+            "THEN ARRAY[ :flightid ] " +
+            "ELSE " +
+            "(SELECT ARRAY_AGG(DISTINCT arrayelt) " +
```

Comment:
I forgot where we landed in our conversation of how closely we wanted to adhere to the SQL standard for this query since we're using arrays. I wasn't sure how I would approach this query with array appends so I looked it up and found multiple interesting approaches including using the `||` operator as outlined in the following StackOverflow answer: https://stackoverflow.com/a/43637976 . It makes the query a bit shorter and I thought it was interesting enough to share. Also interesting is the suggestion to use the `intarray` module bundled with Postgres when dealing with arrays of numbers as it has a `uniq` function (not the case here).

Reply:
Added a comment explaining the logic.

I did consider using the OR operator instead of the ARRAY_APPEND, but decided that the function was easier to follow. Regarding the SQL standard, I don't think that arrays are very consistent across databases, so this exact SQL is not portable regardless of what operators we use.